### PR TITLE
[homematic] Improve stability of resource loading

### PR DESCRIPTION
**Pull request by mdicke2s**
Resolve resources from bundle instead of class loader (see https://github.com/openhab/openhab2-addons/issues/3195) & refactor to try-with-ressources
Additionally: Prevent potential NullPointerException on HomematicDeviceDiscoveryService

Signed-off-by: Michael Reitler <michael.reitler@telekom.de>

### CODE REVIEW

**Comment by mdicke2s** [comment]
@gerrieg Would you mind to take a look?

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java [hunk 1]
```diff
@@ -220,13 +220,23 @@ protected void executeScript(HmDatapoint dp) throws IOException {
...
         Bundle bundle = FrameworkUtil.getBundle(getClass());
-        InputStream stream = bundle.getResource("homematic/tclrega-scripts.xml").openStream();
-        TclScriptList scriptList = (TclScriptList) xStream.fromXML(stream);
-
+        InputStream stream = null;
         Map<String, String> result = new HashMap<String, String>();
-        if (scriptList.getScripts() != null) {
-            for (TclScript script : scriptList.getScripts()) {
-                result.put(script.name, StringUtils.trimToNull(script.data));
+        try {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java]: This is not a try-with-resources, that would look like this:

```java
        try (InputStream stream = bundle.getResource("homematic/tclrega-scripts.xml").openStream()) {
            TclScriptList scriptList = (TclScriptList) xStream.fromXML(stream);

            Map<String, String> result = new HashMap<String, String>();
            if (scriptList.getScripts() != null) {
                for (TclScript script : scriptList.getScripts()) {
                    result.put(script.name, StringUtils.trimToNull(script.data));
                }
            }
            return result;
        }
```

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java [hunk 2]
```diff
@@ -220,13 +220,23 @@ protected void executeScript(HmDatapoint dp) throws IOException {
...
+            stream = bundle.getResource("homematic/tclrega-scripts.xml").openStream();
+            TclScriptList scriptList = (TclScriptList) xStream.fromXML(stream);
+            if (scriptList.getScripts() != null) {
+                for (TclScript script : scriptList.getScripts()) {
+                    result.put(script.name, StringUtils.trimToNull(script.data));
+                }
+            }
+        } catch (NullPointerException | IllegalStateException | IOException e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java]: Please do not catch `NullPointerException`s,

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java [hunk 3]
```diff
@@ -30,11 +33,22 @@
 
     public BatteryTypeVirtualDatapointHandler() {
         try {
-            InputStream is = Thread.currentThread().getContextClassLoader()
-                    .getResourceAsStream("homematic/batteries.properties");
-            batteries.load(is);
+            Bundle bundle = FrameworkUtil.getBundle(getClass());
+            InputStream stream = null;
+            try {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java]: Please use proper try-with-resources

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java [hunk 4]
```diff
@@ -30,11 +33,22 @@
...
+            } catch (NullPointerException | IllegalStateException | IOException e) {
+                throw new IOException("The resource homematic/batteries.properties could not be loaded!", e);
+            } finally {
+                try {
+                    stream.close();
+                } catch (Exception e) {
+                    // nothing to do
+                }
+            }
         } catch (Exception e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java]: This is a bad-practice, but given it was already the case you can leave it.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java [hunk 5]
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
...
-            while ((line = reader.readLine()) != null) {
-                if (StringUtils.trimToNull(line) != null && !StringUtils.startsWith(line, "#")) {
-                    String channelType = StringUtils.trimToNull(StringUtils.substringBefore(line, "|"));
-                    String datapointName = StringUtils.trimToNull(StringUtils.substringAfter(line, "|"));
-
-                    Set<String> channelDatapoints = standardDatapoints.get(channelType);
-                    if (channelDatapoints == null) {
-                        channelDatapoints = new HashSet<String>();
-                        standardDatapoints.put(channelType, channelDatapoints);
+        try {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java]: This was a try-with-resources :-(

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java [hunk 6]
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
...
+            Bundle bundle = FrameworkUtil.getBundle(MetadataUtils.class);
+            InputStream stream = null;
+            BufferedReader reader = null;
+            try {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java]: I see no valid reason to wrap try-catches (here)

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java [hunk 7]
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
...
+                closeQuietly(stream);
             }
-        } catch (IOException ex) {
-            logger.warn("Can't load standard-datapoints.properties file!");
+        } catch (Exception e) {
+            logger.warn("Can't load standard-datapoints.properties file!", e);
+        }
+    }
+
+    private static void closeQuietly(Closeable closable) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java]: You don't need this is if you apply the real try-with-resources

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java [hunk 8]
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
...
                 }
+            } catch (NullPointerException | IllegalStateException | IOException e) {
+                throw new IOException("The resource homematic/standard-datapoints.properties could not be loaded!", e);
+            } finally {
+                closeQuietly(reader);
+                closeQuietly(stream);
             }
-        } catch (IOException ex) {
-            logger.warn("Can't load standard-datapoints.properties file!");
+        } catch (Exception e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java]: Please don't catch `Exception`'s, the `IOException` was fine. Or explain me a good reason to change it.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java [hunk 9]
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
...
+            logger.warn("Can't load standard-datapoints.properties file!", e);
+        }
+    }
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java]: Imho, the correct try-with-resources implementation would be:

```java
    private static void loadStandardDatapoints() {
        Bundle bundle = FrameworkUtil.getBundle(MetadataUtils.class);
        try (InputStream is = bundle.getResource("homematic/standard-datapoints.properties").openStream();
             BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (StringUtils.trimToNull(line) != null && !StringUtils.startsWith(line, "#")) {
                    String channelType = StringUtils.trimToNull(StringUtils.substringBefore(line, "|"));
                    String dataPointName = StringUtils.trimToNull(StringUtils.substringAfter(line, "|"));

                    Set<String> channelDataPoints = standardDatapoints.computeIfAbsent(channelType, k -> new HashSet<>());

                    channelDataPoints.add(dataPointName);
                }
            }
        } catch (IOException ex) {
            logger.warn("Can't load standard-datapoints.properties file!: {}", ex.getMessage());
        }
    }
```

**Comment by mdicke2s** [comment]
@martinvw Thanks for your feedback :+1:

Sorry, that I didn't contribute "try-with-resources" in the first place. I used to have them in my codebase and for some reason I switched back to w/o "try-with-resources". Therefore I updated the PR and changed the following:

- use try-with-resources
- avoid nested try-catch in 2 methods (since the outer catch didn't do anything but logging, we can simply log in the catch block without re-throwing an exception)

**Comment by mdicke2s** [comment]
Great :-)
@martinvw Thank you

**Pull request merged**